### PR TITLE
feat(sdk): add across gas fee estimations

### PR DESCRIPTION
**Pull request by daywiss**
Signed-off-by: David <david@umaproject.org>

**Motivation**

https://app.shortcut.com/uma-project/story/2519/add-across-fast-slow-relay-gas-prices-to-sdk

**Summary**
Adds a gasFeeCalculator file for using to estimate gas fees when relaying. Also does an end to end test with documentation to show how to get the final result. Has estimations for eth/erc20/uma slow/fast transfers.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [x]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

https://app.shortcut.com/uma-project/story/2519/add-across-fast-slow-relay-gas-prices-to-sdk

### CODE REVIEW

File: packages/sdk/src/across/README.md [hunk 1]
```diff
@@ -30,3 +30,41 @@ assert.equal(apyFeePct.toString(), interval.apy)
...
+import * as uma from "@uma/sdk"
+
+const { getGasFees, SLOW_ETH_GAS, SLOW_ERC_GAS, SLOW_UMA_GAS } = uma.across.gasFeeCalculator
+const { fromWei } = uma.across.utils
+
+// given a transfer of ETH across chains, get amount of eth used in gas for a slow transaction
+const gasPrice = await ethersProvider.getGasPrice() // you will need to get an estimate from provider
+const ethFees = getGasfees(SLOW_ETH_GAS, gasPrice)
+
+// given a standard usdc erc20 transfer across chains, get amount of th etoken used in gas for a slow transaction
```
Comment (nicholaspai) [inline packages/sdk/src/across/README.md:53]:
```suggestion
// given a standard usdc erc20 transfer across chains, get amount of the token used in gas for a slow transaction
```

File: packages/sdk/src/across/gasFeeCalculator.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { BigNumber } from "ethers";
+import { ConvertDecimals } from "../utils";
+import { toBNWei, BigNumberish } from "./utils";
+
+export const SLOW_ETH_GAS = 243177;
+export const SLOW_ERC_GAS = 250939;
+export const SLOW_UMA_GAS = 273955;
+
+// fast costs are slightly higher and include the slow cost
```
Comment (nicholaspai) [inline packages/sdk/src/across/gasFeeCalculator.ts:9]: why hard code these? could you not use the [ethers estimateGas](https://docs.ethers.io/v5/api/contract/contract/#contract-estimateGas) feature to make this file more future proof? For instance, whenever Ethereum hard-forks, function gas costs will change as intruction costs change, meaning that these gas limits will not always be accurate.

File: packages/sdk/src/across/gasFeeCalculator.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// returns - total fees in wei
+export const gasToEth = (gas: number, gasPrice: BigNumberish): BigNumber => {
+  return BigNumber.from(gas).mul(ConvertDecimals(9, 18)(gasPrice));
+};
+
+// gas - the gast cost for transfer, use constants defined in file.
+// gasPrice - required, estimated gas price in gwei.
+// price: optional, 1 if not supplied (eth)
+// decimals - optional: number of decimals of token, 18 if not provided
+// returns - the value of fees native to the token/eth provided, in its smallest unit
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/sdk/src/across/gasFeeCalculator.ts:36]: Yeah I like using `@param` and `@return` and `@notice` for example

**Review by nicholaspai — APPROVED** [review]
LGTM but lets definitely change the gas estimation to be dynamic rather than hard coded

File: packages/sdk/src/across/gasFeeCalculator.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (daywiss) [reply packages/sdk/src/across/gasFeeCalculator.ts:36]: yea can do that

**Review by mrice32 — COMMENTED** [review]
This looks great! My main push is that SDK should be doing more for the frontend. I think this should be a black box to the FE that requires a single function call if possible (unless this is undesirable on the FE side).

Thoughts? Curious to hear what @Gamaranto thinks as well.

File: packages/sdk/src/across/utils.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+export const toWei = (num: BigNumberish, decimals?: number) => toBNWei(num, decimals).toString();
+export const fromWei = (num: BigNumberish, decimals?: number) => ethers.utils.formatUnits(num.toString(), decimals);
+
+export type BigNumberish = string | number | BigNumber;
+export type BN = BigNumber;
+
+export function min(a: BigNumberish, b: BigNumberish) {
+  const bna = toBN(a);
+  const bnb = toBN(b);
+  return bna.lte(bnb) ? bna : bnb;
```
[the remark this answers is not in the document]
Comment (daywiss) [reply packages/sdk/src/across/utils.ts:14]: seems like it should but the BigNumber provided with ethers does not have min/max.

**Pull request merged**